src/core/src/k256.rs

```diff
@@ -0,0 +1,105 @@
+use hd_keys_curves_wasm::HDDeriver;
+use k256::elliptic_curve::sec1::{FromEncodedPoint, ToEncodedPoint};
+
+use crate::utils::read_4_byte_be;
+
+pub fn hd_key_derive(data: Vec<u8>) -> Result<Vec<u8>, Vec<u8>> {
+    // 1st arg is a byte for the curve type, 0 is Nist Prime256, 1 is secp256k1
+    // 2nd arg is a 4 byte big-endian integer for the number of bytes in id
+    // 3rd arg is the byte sequence for id
+    // 4th arg is a 4 byte big-endian integer for the number of bytes in cxt
+    // 5th arg is the byte sequence for cxt
+    // 6th arg is a 4 byte big-endian integer for the number of root keys
+    // 7th arg is a variable number of root keys each 64 bytes in length
+    let mut offset = 1;
+
+    let id_len = read_4_byte_be(&data, &mut offset)? as usize;
+    let id = &data[offset..offset + id_len];
+    offset += id_len;
+
+    let cxt_len = read_4_byte_be(&data, &mut offset)? as usize;
+    let cxt = &data[offset..offset + cxt_len];
+    offset += cxt_len;
+
+    let root_keys_len = read_4_byte_be(&data, &mut offset)? as usize;
+    let buffer = &data[offset..];
+
+    if root_keys_len < 2 || (root_keys_len * 33) > buffer.len() {
+        return Err(Vec::new());
+    }
+
+    let mut offset = 0;
+    let mut hd_root_keys = vec![k256::ProjectivePoint::IDENTITY; root_keys_len];
+    for hd_i in hd_root_keys.iter_mut() {
+        let encoded_point = match buffer[offset] {
```

Comment:
We might be able to optimize this because initially we didn't know the preferred format for passing in public keys. We ultimately settled on uncompressed keys with no lead byte which is 64 bytes. If we remove the two other branches here that might reduce the contract size also.

Reply:
not a meaningful gas difference, no worth the changes to working contracts IMO